File: 08-Recursion-Backtracking/02-Subsets-Combinations/java/SubsetsII.java
import java.util.*;
public class SubsetsII {
    public static List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        backtrack(nums, 0, new ArrayList<>(), result);
        return result;
    }
    static void backtrack(int[] nums, int start, List<Integer> cur, List<List<Integer>> result) {
        result.add(new ArrayList<>(cur));
        for (int i = start; i < nums.length; i++) {
            if (i > start && nums[i] == nums[i - 1]) continue;
            cur.add(nums[i]);
            backtrack(nums, i + 1, cur, result);
            cur.remove(cur.size() - 1);
        }
    }
    public static void main(String[] args) { System.out.println(subsetsWithDup(new int[]{1, 2, 2})); }
}

File: 08-Recursion-Backtracking/02-Subsets-Combinations/cpp/subsets_ii.cpp
/*
 * Problem: Subsets II (with duplicates)
 * Source:  LeetCode 90
 * Approach: Sort + backtracking. Skip duplicates at same level.
 * Time: O(n * 2^n), Space: O(n)
 * Tags: backtracking, subsets
 */
#include <bits/stdc++.h>
using namespace std;

void backtrack(vector<int>& nums, int start, vector<int>& cur, vector<vector<int>>& result) {
    result.push_back(cur);
    for (int i = start; i < nums.size(); i++) {
        if (i > start && nums[i] == nums[i - 1]) continue; // skip duplicates
        cur.push_back(nums[i]);
        backtrack(nums, i + 1, cur, result);
        cur.pop_back();
    }
}

int main() {
    vector<int> nums = {1, 2, 2};
    sort(nums.begin(), nums.end());
    vector<vector<int>> result;
    vector<int> cur;
    backtrack(nums, 0, cur, result);
    for (auto& s : result) { cout << "[ "; for (int x : s) cout << x << " "; cout << "]" << endl; }
    return 0;
}
